src/execution/sql/table_vector_iterator.cpp

```diff
@@ -23,41 +31,99 @@ bool TableVectorIterator::Init() {
   return true;
 }
 
 bool TableVectorIterator::Advance() {
   if (!initialized_) return false;
   // First check if the iterator ended.
-  if (*iter_ == table_->end()) {
+  if (*iter_ == *iter_end_) {
     return false;
   }
   // Scan the table to set the projected column.
-  table_->Scan(exec_ctx_->GetTxn(), iter_.get(), projected_columns_);
+  table_->RangeScan(exec_ctx_->GetTxn(), iter_.get(), iter_end_.get(), projected_columns_);
   pci_.SetProjectedColumn(projected_columns_);
   return true;
 }
 
 void TableVectorIterator::Reset() {
   if (!initialized_) return;
-  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->begin());
+  // set the iterator to the start position specified by tbb
+  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->beginAt(start_block_idx_));
 }
 
-bool TableVectorIterator::ParallelScan(uint32_t db_oid, uint32_t table_oid, void *const query_state,
-                                       ThreadStateContainer *const thread_states, const ScanFn scan_fn,
-                                       const uint32_t min_grain_size) {
-  // TODO(Amadou): Implement Me!!
-  return false;
+namespace {
+
+class ScanTask {
+ public:
+  ScanTask(exec::ExecutionContext *exec_ctx, uint16_t table_id, void *const query_state,
+           ThreadStateContainer *const thread_state_container, TableVectorIterator::ScanFn scanner)
+      : exec_ctx_(exec_ctx),
+        table_id_(table_id),
+        query_state_(query_state),
+        thread_state_container_(thread_state_container),
+        scanner_(scanner) {}
+
+  void operator()(const tbb::blocked_range<uint32_t> &block_range) const {
+    // Create the iterator over the specified block range
+    TableVectorIterator iter(exec_ctx_, table_id_, {}, 0, block_range.begin(), block_range.end());
+
+    // Initialize the table vector iterator
+    if (!iter.Init()) {
+      return;
+    }
+
+    // Pull out the thread-local state
+    byte *thread_state = thread_state_container_->AccessThreadStateOfCurrentThread();
+
+    // Call scanning function which should be passed at runtime
+    scanner_(query_state_, thread_state, &iter);
+  }
+
+ private:
+  exec::ExecutionContext *exec_ctx_;
+  uint16_t table_id_;
+  void *const query_state_;
+  ThreadStateContainer *const thread_state_container_;
+  TableVectorIterator::ScanFn scanner_;
+};
+}  // namespace
+
+bool TableVectorIterator::ParallelScan(uint32_t table_oid, void *query_state, ThreadStateContainer *thread_states,
+                                       const ScanFn scan_fn, exec::ExecutionContext *exec_ctx) {
+  // Lookup table
+  common::ManagedPointer<storage::SqlTable> table =
+      exec_ctx->GetAccessor()->GetTable(static_cast<catalog::table_oid_t>(table_oid));
+  if (table == nullptr) {
+    return false;
+  }
+  // get the number of blocks in the table
+  auto block_count = table->GetBlockListSize();
+
+  // TODO(Ron): min_grain_size = num_blocks / num_threads
+  size_t min_grain_size = 3;
```

Comment:
Running with different grain sizes may give more insight in tests, as well

Reply:
No magic constants!

Reply:
We updated the min_grain_size - number of blocks divided by the number of threads we have